tensorflow/python/ops/check_ops.py

```diff
@@ -1504,6 +1505,206 @@ def assert_type(tensor, tf_type, message=None, name=None):
     return control_flow_ops.no_op('statically_determined_correct_type')
 
 
+@tf_export('debugging.assert_shapes', 'assert_shapes')
+def assert_shapes(shapes, event_shape_only=False, data=None, summarize=None,
+                  message=None, name=None):
+  """Assert tensor shapes and dimension size relationships between tensors.
+
+  This Op checks that a collection of tensors shape relationships
+  satisfies given constraints.
+
+  Example:
+
+  ```python
+  tf.assert_shapes({
+    x: ('N', 'Q'),
+    y: ('N', 'D'),
+    param: 'Q',
+    other_param: 2
+  })
+  ```
+
+  If `x`, `y`, `param` or `other_param` does not have a shape that satisfies
+  all specified constraints, `message`, as well as the first `summarize` entries
+  of the first encountered violating tensor are printed, and
+  `InvalidArgumentError` is raised.
+
+  Size entries in the specified shapes are checked explicitly if of type `int`,
+  else checked against other entries by their __hash__.
+
+  Args:
+    shapes: dictionary with (`Tensor` to shape) items. A shape is either a
+    tuple/list of size entries or a single size entry.
+    event_shape_only: `bool`, if to apply constraints on only
```

Comment:
Why is it named event_shape? Also the docstring is somewhat confusing for this method. When would a user want to set it to true vs false?

Reply:
Setting this to true would be useful in scenarios where the user is writing a function that only uses (and have expectations on) the inner-most dimensions of the passed tensors. As when the function is to be broadcasted over any number of the remaining (preceding) dimensions.

An example of such a function is matmul; as where the inner-most two dimensions are to represent the matrix shape of respective operand, while the rest of the dimensions are broadcasted.

I take your point that it was confusing. The name event_shape came from the terminology in 'distributions': https://www.tensorflow.org/api_docs/python/tf/distributions/Normal#event_shape 
But, as this has a wider context, a better name to describe it could be 'innermost_dims', to be in line with terminology as in: https://www.tensorflow.org/api_docs/python/tf/linalg/matmul

However, after having another thought about this flag, the signature I show below could be cleaner and more intuitive. Rather than having a flag at all, this could be achieved meanwhile relying on standard numpy notation for familiar semantics:

Example:
tf.assert_shapes({
	x: (..., 'a', 'b'),
	y: (..., 'b', 'c')
})

Here the rule applied is that if the specified shape contains `...` as the first element, the remaining elements are checked as inner-most (relative to the end via negative indexing). I.e. the first dimensions up to the specified ones are ignored, similar to how `...` is used in numpy. Specifying `...` in any other entry is disallowed (raising a ValueError) to be in line with that semantic and allow for error messages pointing to specific dimensions (as e.g. 'dimension 0 must have size 3' or 'dimension -2 must have size 3).

What do you think?

Reply:
I love this.

Reply:
Great!